### PR TITLE
Implement dependency resolution through package registry

### PR DESCRIPTION
**Pull request by mattt**
This PR builds on top of https://github.com/apple/swift-package-manager/pull/3635 to implement dependency resolution through a registry service.

### Motivation:

Implement [SE-0292](https://github.com/apple/swift-evolution/blob/main/proposals/0292-package-registry-service.md).

### Modifications:

- [x] Add `PackageRegistry` library and test target containing the `RegistryManager` client implementation.
- [x] Add `RegistryPackageContainer` class to `PackageGraph` module
- [x] Update `Workspace` to provide `RegistryPackageContainer` if the package is declared with a registry identifier (**Currently WIP with hardcoded registry URL**)
- [x] Update `PubGrubDependencyResolver` to support registry packages

### Result:

With this change, Swift Package Manager will be able to resolve package dependencies through a configured registry service.

### CODE REVIEW

File: Sources/PackageGraph/RepositoryPackageContainer.swift [hunk 1]
```diff
@@ -325,88 +325,9 @@ public class RepositoryPackageContainer: PackageContainer, CustomStringConvertib
     }
 }
 
-/// Adaptor for exposing repositories as PackageContainerProvider instances.
-///
-/// This is the root class for bridging the manifest & SCM systems into the
-/// interfaces used by the `DependencyResolver` algorithm.
-public class RepositoryPackageContainerProvider: PackageContainerProvider {
```
Comment (mattt) [inline Sources/PackageGraph/RepositoryPackageContainer.swift:332]: @tomerd @abertelrud After digging into the implementation a bit more, I hit a fork in the road and was interested to hear your ideas for how best to proceed.

Currently, responsibility for creating package containers falls to `RepositoryPackageContainerProvider`, which conforms to the `PackageContainerProvider` protocol. 

My first instinct was to either create a complement to `RepositoryPackageContainerProvider` for registry packages, or to update that class to handle both repository and registry packages. However, I noticed that `RepositoryPackageContainerProvider` doesn't do much beyond copying some state from a `Workspace` and implementing the `getContainer` method requirement.

So my thinking here was that we could remove a layer of indirection by making `Workspace` conform to `PackageContainerProvider` directly. This way, we can access all of the information we need directly, (like registry configuration, netrc credentials, etc.).

Do you think this is the right approach to take?

**Comment by tomerd** [comment]
hi @mattt looks like both of us made similar changes to managed deps. my version is https://github.com/apple/swift-package-manager/pull/3698 which I need to get over the finish line. the goal with that PR was to help with SE-0292 implementation so its good that the changes somewhat align. how do we best consolidate?

**Comment by mattt** [comment]
@tomerd Glad to see that we were thinking pretty much the same thing for `ManagedDependency`. I've rebased this PR against #3698 and opened a new PR with additional refactoring ideas in #3750.

**Comment by mattt** [comment]
@tomerd I just finished rebasing this against your changes in #3766. Still some cleanup to do on my end, but resolution through a registry is working as of 35d9d9e.

**Comment by mattt** [comment]
@swift-ci Please smoke test

**Comment by mattt** [comment]
Back to draft to rebase on latest of `main`

**Comment by tomerd** [comment]
replaced by #3863  #3864 #3873 #3874

**Pull request closed without merging**